### PR TITLE
[vtctld] Migrate topo management RPCs

### LINKED ISSUE #7403: Migrate topo management commands to VtctldServer
#### Feature Description

Migrate the following vtctl commands to the new vtctld server:
- [ ] `ChangeTabletType`
- [ ] `CreateKeyspace`
- [ ] `CreateShard`
- [ ] `DeleteKeyspace`
- [ ] `DeleteShard`
- [ ] `DeleteTablet`
- [ ] `GetShard`
- [ ] `RemoveKeyspaceCell`
- [ ] `RemoveShardCell`

### PR DESCRIPTION
**Pull request by ajm188**
### Preface

Okay this is a large PR because I couldn't figure out a good way to add RPCs to the proto file in parallel branches without conflicting with those same other branches; in the future (or I can take an afternoon to restructure commits on this one) I could just add a bunch of RPC definitions that panic when called in a single upfront PR and then implement each in a parallel branch. But anyway, onward!

The branch is structured in a series of paired commits, mostly. It goes essentially like:

```
commit - Add all RPC definitions
foreach rpc
    commit - implement server code and tests
    commit - add CLI entrypoint
end
```

Occasionally there's an extra commit or two when I needed to touch code outside of `grpcvtctldserver` and `cmd/vtctldclient`, but you can mostly view the changes one command at a time by looking at those pairs of commits.

<hr/>

This PR adds a slew of topo-management RPCs to the vtctld server, namely:
- `ChangeTabletType`
- `CreateKeyspace`
- `CreateShard`
- `DeleteKeyspace`
- `DeleteShards` (in the old vtctl server, this was called `DeleteShard` but it could delete multiple shards)
- `DeleteTablets` (same renaming as above)
- `GetShard`
- `RemoveKeyspaceCell`
- `RemoveShardCell`

In addition, I made some refactors to other parts of vitess to facilitate this change, in particular:
- Added a helper `KeyspaceIDTypeString` to `package key`, similar to `topoproto.TabletAliasString` but for the `topodatapb.KeyspaceIdType` enum.
- Added another helper `KeyspaceTypeString` to `package topoproto`, same as above but for the `topodatapb.KeyspaceType` enum. 
    - Note on above 2: I don't know why `package key` had helpers for some of the stuff in `topodatapb`, instead of it all being in `topoproto`, but all the `KeyspaceIdType` helpers were there, so that's why I added it to `package key` instead of `package topoproto`. Happy to just move it if folks feel otherwise.
- Moved the private `isMasterTablet` from `wrangler` to `topotools.IsPrimaryTablet` so I could reuse the code in both.
- Extracted an argument parser helper function from `vtctl` to `package topo`. I wanted to put this in `topoproto`, but because `topo` imports `topoproto`, and this function needs to import `topo`, this would have created a circular import.
### See it in action

I exercised all the new commands against the local examples, and will include their output here:

<details>
<summary><code>ChangeTabletType</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-24T15:42:50Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
❯ vtctldclient --server "localhost:15999" ChangeTabletType zone1-0000000102 replica
- zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
+ zone1-0000000102 commerce 0 replica SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-24T15:42:50Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 replica SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
```

</details>

<details>
<summary><code>CreateKeyspace</code></summary>

 ```
❯ vtctldclient --server "localhost:15999" CreateKeyspace testkeyspace
Successfully created keyspace testkeyspace. Result:
{
  "name": "testkeyspace",
  "keyspace": {
    "sharding_column_name": "",
    "sharding_column_type": "UNSET",
    "served_froms": [
    ],
    "keyspace_type": "NORMAL",
    "base_keyspace": "",
    "snapshot_time": null
  }
}
```

```
❯ vtctldclient --server "localhost:15999" CreateKeyspace --type snapshot --base-keyspace testkeyspace --snapshot-timestamp "2020-12-31T01:02:03.00Z" testsnapshot
Successfully created keyspace testsnapshot. Result:
{
  "name": "testsnapshot",
  "keyspace": {
    "sharding_column_name": "",
    "sharding_column_type": "UNSET",
    "served_froms": [
    ],
    "keyspace_type": "SNAPSHOT",
    "base_keyspace": "testkeyspace",
    "snapshot_time": {
      "seconds": "1609376523",
      "nanoseconds": 0
    }
  }
}
```

```
❯ vtctldclient --server "localhost:15999" CreateKeyspace testkeyspace2 --served-from "master:testkeyspace,rdonly:commerce"
Successfully created keyspace testkeyspace2. Result:
{
  "name": "testkeyspace2",
  "keyspace": {
    "sharding_column_name": "",
    "sharding_column_type": "UNSET",
    "served_froms": [
      {
        "tablet_type": "MASTER",
        "cells": [
        ],
        "keyspace": "testkeyspace"
      },
      {
        "tablet_type": "RDONLY",
        "cells": [
        ],
        "keyspace": "commerce"
      }
    ],
    "keyspace_type": "NORMAL",
    "base_keyspace": "",
    "snapshot_time": null
  }
}
```

</details>

<details>
<summary><code>GetShard</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetShard commerce/-
E0125 11:03:51.136664   14959 main.go:42] rpc error: code = Unknown desc = node doesn't exist: /vitess/global/keyspaces/commerce/shards/-/Shard
❯ vtctldclient --server "localhost:15999" GetShard commerce/0
{
  "keyspace": "commerce",
  "name": "0",
  "shard": {
    "master_alias": {
      "cell": "zone1",
      "uid": 100
    },
    "master_term_start_time": {
      "seconds": "1611590612",
      "nanoseconds": 941055000
    },
    "key_range": null,
    "served_types": [
    ],
    "source_shards": [
    ],
    "tablet_controls": [
    ],
    "is_master_serving": true
  }
}
```

</details>

<details>
<summary><code>RemoveShardCell</code></summary>

```
❯ ./101_initial_cluster.sh
❯ ./201_customer_tablets.sh
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-25T23:15:11Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
zone1-0000000200 customer 0 master SFO-M-AMASON02:15200 SFO-M-AMASON02:17200 [] 2021-01-25T23:21:39Z
zone1-0000000201 customer 0 replica SFO-M-AMASON02:15201 SFO-M-AMASON02:17201 [] <null>
zone1-0000000202 customer 0 rdonly SFO-M-AMASON02:15202 SFO-M-AMASON02:17202 [] <null>
❯ vtctlclient -server "localhost:15999" DeleteTablet -allow_master zone1-0000000200
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-25T23:15:11Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
zone1-0000000201 customer 0 replica SFO-M-AMASON02:15201 SFO-M-AMASON02:17201 [] <null>
zone1-0000000202 customer 0 rdonly SFO-M-AMASON02:15202 SFO-M-AMASON02:17202 [] <null>
❯ vtctldclient --server "localhost:15999" RemoveShardCell customer/0 zone1
E0125 18:22:29.837635   40277 main.go:42] rpc error: code = Unknown desc = cell zone1 has 2 possible tablets in replication graph
❯ vtctldclient --server "localhost:15999" RemoveShardCell customer/0 zone1 --recursive
Successfully removed cell zone1 from shard customer/0
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-25T23:15:11Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
```

</details>

<details>
<summary><code>RemoveKeyspaceCell</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-26T01:28:23Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
zone1-0000000200 customer 0 master SFO-M-AMASON02:15200 SFO-M-AMASON02:17200 [] 2021-01-26T01:29:14Z
zone1-0000000201 customer 0 replica SFO-M-AMASON02:15201 SFO-M-AMASON02:17201 [] <null>
zone1-0000000202 customer 0 rdonly SFO-M-AMASON02:15202 SFO-M-AMASON02:17202 [] <null>
❯ vtctlclient -server "localhost:15999" deletetablet -allow_master zone1-0000000200
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-26T01:28:23Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
zone1-0000000201 customer 0 replica SFO-M-AMASON02:15201 SFO-M-AMASON02:17201 [] <null>
zone1-0000000202 customer 0 rdonly SFO-M-AMASON02:15202 SFO-M-AMASON02:17202 [] <null>
❯ vtctldclient --server "localhost:15999" RemoveKeyspaceCell customer zone1 --recursive
Successfully removed keyspace customer from cell zone1
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-26T01:28:23Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
```

</details>

<details>
<summary><code>CreateShard</code></summary>

```
❯ vtctldclient --server "localhost:15999" CreateShard commerce/-40
{
  "keyspace": {
    "name": "commerce",
    "keyspace": {
      "sharding_column_name": "",
      "sharding_column_type": "UNSET",
      "served_froms": [
      ],
      "keyspace_type": "NORMAL",
      "base_keyspace": "",
      "snapshot_time": null
    }
  },
  "shard": {
    "keyspace": "commerce",
    "name": "-40",
    "shard": {
      "master_alias": null,
      "master_term_start_time": null,
      "key_range": {
        "start": null,
        "end": "QA=="
      },
      "served_types": [
      ],
      "source_shards": [
      ],
      "tablet_controls": [
      ],
      "is_master_serving": false
    }
  },
  "shard_already_exists": false
}
❯ vtctldclient --server "localhost:15999" CreateShard testkeyspace/- -p
{
  "keyspace": {
    "name": "testkeyspace",
    "keyspace": {
      "sharding_column_name": "",
      "sharding_column_type": "UNSET",
      "served_froms": [
      ],
      "keyspace_type": "NORMAL",
      "base_keyspace": "",
      "snapshot_time": null
    }
  },
  "shard": {
    "keyspace": "testkeyspace",
    "name": "-",
    "shard": {
      "master_alias": null,
      "master_term_start_time": null,
      "key_range": {
        "start": null,
        "end": null
      },
      "served_types": [
      ],
      "source_shards": [
      ],
      "tablet_controls": [
      ],
      "is_master_serving": true
    }
  },
  "shard_already_exists": false
}
❯ vtctldclient --server "localhost:15999" CreateShard testkeyspace/- -pf
{
  "keyspace": {
    "name": "testkeyspace",
    "keyspace": {
      "sharding_column_name": "",
      "sharding_column_type": "UNSET",
      "served_froms": [
      ],
      "keyspace_type": "NORMAL",
      "base_keyspace": "",
      "snapshot_time": null
    }
  },
  "shard": {
    "keyspace": "testkeyspace",
    "name": "-",
    "shard": {
      "master_alias": null,
      "master_term_start_time": null,
      "key_range": {
        "start": null,
        "end": null
      },
      "served_types": [
      ],
      "source_shards": [
      ],
      "tablet_controls": [
      ],
      "is_master_serving": true
    }
  },
  "shard_already_exists": true
}
```

</details>

<details>
<summary><code>DeleteTablets</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-26T14:54:12Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
❯ vtctldclient --server "localhost:15999" DeleteTablets zone1-0000000101 zone1-0000000102
Successfully deleted 2 tablets
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-26T14:54:12Z
❯ vtctldclient --server "localhost:15999" DeleteTablets zone1-0000000100
Error: rpc error: code = Unknown desc = cannot delete tablet zone1-0000000100 as it is a master, pass AllowPrimary = true: while deleting 1 tablets; please inspect the topo
❯ vtctldclient --server "localhost:15999" DeleteTablets zone1-0000000100 -p
Successfully deleted 1 tablets
❯ vtctldclient --server "localhost:15999" GetTablets
```

</details>

<details>
<summary><code>DeleteShards</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-27T03:14:46Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
❯ vtctldclient --server "localhost:15999" DeleteShards -rf commerce/0
Successfully deleted 1 shards
❯ vtctldclient --server "localhost:15999" GetTablets
```

</details>

<details>
<summary><code>DeleteKeyspace</code></summary>

```
❯ vtctldclient --server "localhost:15999" GetTablets
zone1-0000000100 commerce 0 master SFO-M-AMASON02:15100 SFO-M-AMASON02:17100 [] 2021-01-27T12:57:20Z
zone1-0000000101 commerce 0 replica SFO-M-AMASON02:15101 SFO-M-AMASON02:17101 [] <null>
zone1-0000000102 commerce 0 rdonly SFO-M-AMASON02:15102 SFO-M-AMASON02:17102 [] <null>
❯ vtctldclient --server "localhost:15999" GetKeyspaces
[
  {
    "name": "commerce",
    "keyspace": {}
  }
]
❯ vtctldclient --server "localhost:15999" DeleteKeyspace commerce
Error: DeleteKeyspace(commerce) error: rpc error: code = Unknown desc = keyspace commerce still has 1 shards; use Recursive=true or remove them manually; please check the topo
❯ vtctldclient --server "localhost:15999" DeleteKeyspace -r commerce
Successfully deleted keyspace commerce.
❯ vtctldclient --server "localhost:15999" GetKeyspaces
null
❯ vtctldclient --server "localhost:15999" GetTablets
```

</details>

## Related Issue(s)

- #7058 
- #7135 
- This closes #7403 

## Checklist
- [ ] Should this PR be backported? **No**
- [x] Tests were added or are not required _oh boy were they ever_
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

File: proto/vtctldata.proto
Comment (doeg) [inline proto/vtctldata.proto:58]: Just curious -- [CreateKeyspace docs](https://vitess.io/docs/reference/programs/vtctl/keyspaces/#createkeyspace) don't mention a few of these params. Do you think this is usual doc drift or are there $reasons they aren't included? (Or am I looking at the wrong thing entirely?)

File: proto/vtctldata.proto
Comment (doeg) [inline proto/vtctldata.proto:74]: On `served_froms` specifically, the docs list `served_from` 🤔 are they supposed to be the same?

**Review by doeg — COMMENTED** [review]
I'm still reviewing this! Had a couple of preliminary questions to make sure I'm interpreting this vs. the docs correctly. :)

File: proto/vtctldata.proto
Comment (ajm188) [reply proto/vtctldata.proto:58]: Yeah, my broad answer on this is that these are similar to but distinctly **not** the vtctl commands, and that includes the CLI definitions (and, those docs don't actually list [all the flags you can actually use](https://github.com/vitessio/vitess/blob/4fea10e8718e4285f1a1c2675cb85ed458ba4565/go/vt/vtctl/vtctl.go#L1619-L1628) in the current version). So exact matching is not a requirement, and I figured this was a good opportunity to improve on the existing CLI options rather than just carry them over.

**Review by doeg — COMMENTED** [review]
Ok I read this whole thing and, first of all: this is amazing. An immense amount of work + as usual you are so _very_ thorough.

Standard caveat I am but a golang/Vitess novice, so definitely get a review from a Vitess-knower! (I'm going to comment instead of approve for that reason.) :) 

From my perspective though, this is all very approachable and easy to follow. Your comments + tests are 🏆 so good. 🏆  It is a pleasure to learn from you. 😎

File: go/vt/key/key.go
Comment (ajm188) [reply go/vt/key/key.go:198]: githook linter caught it for me 😀

File: go/cmd/vtctldclient/internal/command/keyspaces.go
Comment (ajm188) [reply go/cmd/vtctldclient/internal/command/keyspaces.go:96]: Yep, those are exactly equivalent! I think also you could remove the `kst` variable and call the function both times, and the compiler _should_ (🤞) be smart enough to simplify the expression and store the result of the first call.

> If there are other advantages over an if, I'm curious to know them!

I think it's mainly a style convention! I've come to find it more readable, and I _think_ it's generally preferred in Go to write `switch` statements over long `if-else-if-else` chains, which is hinted at, but not explicitly stated in [effective go](https://golang.org/doc/effective_go.html#switch).

File: go/vt/topotools/tablet.go
Comment (ajm188) [reply go/vt/topotools/tablet.go:106]: Thanks! 😊

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (setassociative) [inline go/vt/vtctl/grpcvtctldserver/server.go:96]: I'd at least catch the error and log ... something? getting a Nil back for after tablet on success is a bit weird and we'll want some kind trail if we hit it

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (setassociative) [inline go/vt/vtctl/grpcvtctldserver/server.go:106]: I understand these implementations are intended as a leg over from having the implementation tied up in `vtctl/vtctl.go::commandCreateKeyspace` etc but it feels odd that many of these implementations are so heavily duplicated and could result in drift. How do we plan on preventing that (rapid deprecation+removal of legacy impl is a totally valid answer).

Even so one thing I'd like us to consider is how we might want to detach the command functionality from the transport? E.g. what does it look like if we also wanted a REST API or something that kicked off create keyspace etc. (Viable answer here is : we only ever care about grpc but _if that changes_ the req/resp objects provide a clear way to introduce a layer of indirection so we don't have to duplicate code again).

**Review by setassociative — COMMENTED** [review]
I want to sit down with a real monitor and A/B the new implementations from the ones they're based on but I don't expect anything surprising. I had a few nit picks but nothing that is actually concerning or suggests major changes, I think.

File: proto/vtctldata.proto
Comment (setassociative) [reply proto/vtctldata.proto:74]: I think I can take this one -- the existing docs list the arg as `-served_from=tablettype1:ks1,tablettype2:ks2,...`

but served_from takes a list and gets mapped into ServedFroms for the actual call if more than one is provided:

```go
        if len(servedFrom) > 0 {
                for name, value := range servedFrom {
                        tt, err := parseServingTabletType3(name)
                        if err != nil {
                                return err
                        }
                        ki.ServedFroms = append(ki.ServedFroms, &topodatapb.Keyspace_ServedFrom{
                                TabletType: tt,
                                Keyspace:   value,
                        })
                }
        }
```

This is formalizing that CLI behavior into a request object.

File: go/cmd/vtctldclient/internal/command/keyspaces.go
Comment (ajm188) [reply go/cmd/vtctldclient/internal/command/keyspaces.go:169]: That's a great point. I started down the following compromise:

```go
resp, err := client.DeleteKeyspace(...)
if err != nil {
    return fmt.Errorf("...", ks, err)
}

defer func() { fmt.Fprintf(os.Stderr, "Successfully deleted keyspace %v.\n", ks }()

data, err := cli.MarshalJSON(resp)
if err != nil {
    return err
}

fmt.Printf("%s\n", data)

return nil
```

Which produces:

```
❯ vtctldclient --server "localhost:15999" DeleteKeyspace -r commerce
{

}
Successfully deleted keyspace commerce.
```

And I _kinda_ wish I got back bytes that let me do something like `if len(data) > 0 { fmt.Printf("%s\n", data) }` but I'm also not sure how to do that without losing the pretty printing (and also the curly braces for the empty-but-non-null object).

What we _could_ do now that I'm thinking this through is expand `cli.MarshalJSON`'s `proto.Message` case to `protoreflect` on the message and return an empty slice if there are no real fields in the message. However. That only works if we're using `google.golang.org/protobuf/proto`, which provides a `protoreflect` package that works with the `proto.Message` type defined in that module, but does not work with the types in `github.com/golang/protobuf/proto`, because the former is technically a v2 of the latter and they didn't keep backwards compatibility on this one.

So I think either of these has to be Good Enough, but I'm also curious if you have other ideas after reading all this:
1. The "Really Hope We Don't Forget" approach. We could also add something to the PR template, but that feels a bit heavy-handed since it will not apply to something like 99% of PRs.
2. Accept that these commands will print an empty object over 3 lines (`[]byte("{\n\n}")`, plus the newline from `fmt.Printf("%s\n", data)`)

File: go/vt/topotools/tablet.go
Comment (ajm188) [reply go/vt/topotools/tablet.go:117]: This is an extraction of [`isMasterTablet`](https://github.com/vitessio/vitess/blob/master/go/vt/wrangler/tablet.go#L222-L251), which now calls this

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [reply go/vt/vtctl/grpcvtctldserver/server.go:106]: > but it feels odd that many of these implementations are so heavily duplicated and could result in drift. How do we plan on preventing that (rapid deprecation+removal of legacy impl is a totally valid answer).

Totally. I've stuck a [`VtctldServer` inside of the wrangler](https://github.com/vitessio/vitess/blob/master/go/vt/wrangler/wrangler.go#L48), which powers the legacy implementation. My plan here is to make another pass after this PR and replace all the logic with a call to `wr.VtctldServer().CreateKeyspace(...)` and so on, and then munge those responses back into the legacy output format. 

I didn't do this in this PR mainly because this branch was already gigantic, and it'll let me just focus on making sure I don't accidentally change the legacy output, which I've [already done once](https://github.com/vitessio/vitess/pull/7285).

> one thing I'd like us to consider is how we might want to detach the command functionality from the transport?

This one is really interesting. My answer for the legacy CLI is still to deprecate+remove. But, more abstractly about the REST case:

I'm not sure how we can guarantee client/server type safety without adding something like jsonschema to the mix, but ignoring that, I would probably use the grpc api as the de facto standard, and add adapters for REST/whatever, so we would:

1. Add a global flag to `vtctldclient` that says which transport to use. Add any extra flags needed to set up those transports (like the `grpc_static_auth_creds` flag for the grpc transport)
2. Create `package httpvtctldclient` in the same style as `package grpcvtctldclient`. This is where we'll translate the proto message objects into an HTTP request, and then the HTTP response back into a proto message. Similar for any other transports we want to support.
3. In `Root.PersistentPreRunE`, instantiate the correct client (`grpcvtctldclient.New(...)`, `httpvtctldclient.New(...)`, etc), everything about the individual command implementations is unchanged.

File: go/vt/vtctl/grpcvtctldserver/server.go
Comment (ajm188) [reply go/vt/vtctl/grpcvtctldserver/server.go]: ✅

File: go/cmd/vtctldclient/internal/command/keyspaces.go
Comment (ajm188) [reply go/cmd/vtctldclient/internal/command/keyspaces.go]: ✅

File: go/vt/topotools/tablet.go
Comment (deepthi) [reply go/vt/topotools/tablet.go:117]: @setassociative we went through those scenarios while writing the original version, so it should be fine.
I think the concern about public vs private function is valid, but that can be addressed in the PR that replaces the legacy command implementations with calls to `VtctldServer()....`.

**Review by rohit-nayak-ps — COMMENTED** [review]
lgtm
a couple of trivial nits. will merge once addressed

File: go/cmd/vtctldclient/cli/pflag.go
Comment (ajm188) [reply go/cmd/vtctldclient/cli/pflag.go:42]: [Here](https://github.com/vitessio/vitess/pull/7395/files#diff-770a0519560ac868354fb300c9156766248c8405724c28704888eb58e2bf581aR82)

File: go/cmd/vtctldclient/cli/pflag.go
Comment (ajm188) [reply go/cmd/vtctldclient/cli/pflag.go:29]: ✅

File: go/cmd/vtctldclient/cli/shards.go
Comment (ajm188) [reply go/cmd/vtctldclient/cli/shards.go:2]: ✅

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

**Pull request merged**
